Enable parallel planning with PipelinePlanner

sjahr opened the pull request:
This PR refactors the PipelinePlanner to enable the use of MoveIt's parallel planning API. Usage would look for example like this

```cpp
  // Configure a set of pipelines to be used
  std::unordered_map<std::string, std::string> pipeline_id_planner_id_map;
  pipeline_id_planner_id_map["ompl"] = "RRTConnectkConfigDefault";
  pipeline_id_planner_id_map["pilz_industrial_motion_planner"] = "LIN";
  pipeline_id_planner_id_map["chomp"] = "chomp";

  // Construct solver
  auto parallel_planner = std::make_shared<mtc::solvers::PipelinePlanner>(node_, pipeline_id_planner_id_map);
  
  // Optional: Configure custom solution selection and/or stopping criterion
  parallel_planner->setStoppingCriterionFunction(...);
  parallel_planner->setSolutionSelectionFunction(
      [](const std::vector<::planning_interface::MotionPlanResponse>& solutions) {...});
```

You can test the PR by using [my fork of the moveit2_tutorials](https://github.com/sjahr/moveit2_tutorials/tree/update_parallel_planning_tutorial). Build the custom MTC and moveit2_tutorials branch and run in two terminal windows:
 
1. `ros2 launch moveit2_tutorials mtc_demo.launch.py`
2. `ros2 launch moveit2_tutorials pick_place_demo.launch.py`

rhaschke commented:
Have a look at #429, which provides parallel planning functionality not only including pipeline planners but also joint interpolation and Cartesian interpolation.

sjahr commented:
> Have a look at https://github.com/ros-planning/moveit_task_constructor/pull/429, which provides parallel planning functionality not only including pipeline planners but also joint interpolation and Cartesian interpolation.

Cool! I was not aware of this. As I understand it, it is still different from the proposed Parallel Planning solver as you can not customize stopping and solution selection criteria and the MultiPlanner is capable of using multiple MTC solvers at the same time. So I'd see both as alternatives.

rhaschke commented:
That's true. The MultiPlanner defines a _sequence_ of planning alternatives that are tried _one after the other_, while parallel planning considers all alternatives in parallel. Would it be possible to apply the same approach as in MultiPlanner to realize your goal? This would allow to include joint-space and Cartesian-space interpolation planners as alternatives as well.

By the way, a stopping criterium might be useful for the existing sequential MultiPlanner as well - to continue planning if the criterium (e.g. smoothness) is not yet satisfied.

What about the _new_ names `FallbacksPlanner` and `AlternativesPlanner` to emphasize the similarity to the corresponding parallel containers?

sjahr commented:
> That's true. The MultiPlanner defines a _sequence_ of planning alternatives that are tried _one after the other_, while parallel planning considers all alternatives in parallel. Would it be possible to apply the same approach as in MultiPlanner to realize your goal? This would allow to include joint-space and Cartesian-space interpolation planners as alternatives as well.
> 
> By the way, a stopping criterium might be useful for the existing sequential MultiPlanner as well - to continue planning if the criterium (e.g. smoothness) is not yet satisfied.
> 
> What about the _new_ names `FallbacksPlanner` and `AlternativesPlanner` to emphasize the similarity to the corresponding parallel containers?

That's a good idea :+1: I'll do that and use this chance to port the MultiPlanner over to the ros2 branch and rename it

sea-bass commented on core/src/solvers/pipeline_planner.cpp:
```diff
@@ -49,164 +49,172 @@
+		pipeline_id_planner_id_map_[pipeline_name] = planner_id;
 	}
-	return planner;
+	RCLCPP_WARN(node_->get_logger(), "PipelinePlanner does not have a pipeline called '%s'. Cannot set pipeline ID '%s'",
+	            pipeline_name.c_str(), planner_id.c_str());
```
Should this be an `RCLCPP_ERROR`?

sea-bass commented on core/include/moveit/task_constructor/solvers/pipeline_planner.h:
```diff
@@ -56,48 +59,131 @@ MOVEIT_CLASS_FORWARD(PipelinePlanner);
+	 * \param [in] stopping_criterion_callback Callback function to stop parallel planning pipeline according to a user defined criteria
+	 * \param [in] solution_selection_function Callback function to choose the best solution when multiple pipelines are used
+	 */
+	PipelinePlanner(const rclcpp::Node::SharedPtr& node,
+	                const std::unordered_map<std::string, std::string>& pipeline_id_planner_id_map,
+	                const std::unordered_map<std::string, planning_pipeline::PlanningPipelinePtr>& planning_pipelines =
+	                    std::unordered_map<std::string, planning_pipeline::PlanningPipelinePtr>(),
+						const moveit::planning_pipeline_interfaces::StoppingCriterionFunction& stopping_criterion_callback = &moveit::planning_pipeline_interfaces::stopAtFirstSolution,
+    const moveit::planning_pipeline_interfaces::SolutionSelectionFunction& solution_selection_function = &moveit::planning_pipeline_interfaces::getShortestSolution);
```
Did auto-formatter do this? The indentation in the last 3 lines is very weird.

sea-bass reviewed (commented):
Looking good! I have a few minor comments.

I tried your tutorial fork and it works, but the placement pose doesn't seem correct. However, also seems to look like that on `main`, so I don't think it's a problem with this branch.

https://user-images.githubusercontent.com/4603398/231763209-c76e7ee1-d2ec-49a1-a424-4a855e0eace3.mp4

rhaschke commented:
As stated in https://github.com/ros-planning/moveit_task_constructor/pull/451#issuecomment-1505410218, I don't see added value in this PR in addition to #451. Before you put more effort into this, we should agree on the general route...

sjahr commented:
> As stated in [#451 (comment)](https://github.com/ros-planning/moveit_task_constructor/pull/451#issuecomment-1505410218), I don't see added value in this PR in addition to #451. Before you put more effort into this, we should agree on the general route...

Yes :+1: I answered in the other thread

MarqRazz commented:
Is this actually planning in parallel? The console output looks suspiciously in series, but maybe that's just the ROS logging framework.
```
[mtc_tutorial-1] [INFO] [1681497796.624044521] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'panda_arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[mtc_tutorial-1] [INFO] [1681497796.661303311] [moveit.::planning_interface.planning_pipeline_interfaces]: Stopping criterion met: Terminating planning pipelines that are still active
[mtc_tutorial-1] [INFO] [1681497796.684124377] [moveit.pilz_industrial_motion_planner.trajectory_generator]: Generating LIN trajectory...
[mtc_tutorial-1] [INFO] [1681497796.684154697] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'panda_arm[RRTConnectkConfigDefault]' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[mtc_tutorial-1] [ERROR] [1681497796.684650693] [moveit.pilz_industrial_motion_planner.trajectory_functions]: Joint acceleration limit of panda_joint6 violated. Set the acceleration scaling factor lower! Actual joint acceleration is 5.1937, while the limit is 5. 
[mtc_tutorial-1] [ERROR] [1681497796.684675490] [moveit.pilz_industrial_motion_planner.trajectory_functions]: Inverse kinematics solution at 0.2s violates the joint velocity/acceleration/deceleration limits.
[mtc_tutorial-1] [ERROR] [1681497796.684719357] [moveit.pilz_industrial_motion_planner.trajectory_generator]: Failed to generate valid joint trajectory from the Cartesian path
[mtc_tutorial-1] [INFO] [1681497796.831316182] [chomp_planner]: CHOMP trajectory initialized using method: quintic-spline
[mtc_tutorial-1] [INFO] [1681497796.831353964] [chomp_optimizer]: Active collision detector is: HYBRID
[mtc_tutorial-1] [INFO] [1681497796.885109117] [chomp_optimizer]: First coll check took 0.053745 sec
[mtc_tutorial-1] [INFO] [1681497796.897210374] [chomp_optimizer]: Chomp Got mesh to mesh safety at iter 0. Breaking out early.
[mtc_tutorial-1] [INFO] [1681497796.897222659] [chomp_optimizer]: Chomp path is collision free
[mtc_tutorial-1] [INFO] [1681497796.897225003] [chomp_optimizer]: Terminated after 1 iterations, using path from iteration 0
[mtc_tutorial-1] [INFO] [1681497796.897226478] [chomp_optimizer]: Optimization core finished in 0.004804 sec
[mtc_tutorial-1] [INFO] [1681497796.897228978] [chomp_optimizer]: Time per iteration 0.004806 sec
[mtc_tutorial-1] [INFO] [1681497798.548378078] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'panda_arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[mtc_tutorial-1] [INFO] [1681497798.589949033] [moveit.::planning_interface.planning_pipeline_interfaces]: Stopping criterion met: Terminating planning pipelines that are still active
[mtc_tutorial-1] [INFO] [1681497798.726495083] [moveit.pilz_industrial_motion_planner.trajectory_generator]: Generating LIN trajectory...
[mtc_tutorial-1] [INFO] [1681497798.726616093] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'panda_arm[RRTConnectkConfigDefault]' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[mtc_tutorial-1] Length of solution for planner: 'LIN' is -1'
[mtc_tutorial-1] Length of solution for planner: 'RRTConnectkConfigDefault' is 6.14397'
[mtc_tutorial-1] Length of solution for planner: 'chomp' is 6.14404'
```

Why are we getting the message
```
Stopping criterion met: Terminating planning pipelines that are still active
```
when we have `Omitted default stopping criterion and allow all pipelines to use the full budget`?

sjahr commented:
> Is this actually planning in parallel?

It's probably the ROS logging. Look at the first two lines, PILZ and OMPL logging are interwoven which wouldn't be the case if it is sequential:

```bash
[mtc_tutorial-1] [INFO] [1681497796.624044521] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'panda_arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[mtc_tutorial-1] [INFO] [1681497796.661303311] [moveit.::planning_interface.planning_pipeline_interfaces]: Stopping criterion met: Terminating planning pipelines that are still active
[mtc_tutorial-1] [INFO] [1681497796.684124377] [moveit.pilz_industrial_motion_planner.trajectory_generator]: Generating LIN trajectory...
[mtc_tutorial-1] [INFO] [1681497796.684154697] [moveit.ompl_planning.model_based_planning_context]: Planner configuration 'panda_arm[RRTConnectkConfigDefault]' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
```

> Why are we getting the message

I fixed that, the parallel_planning_solver worked fine but the sampling_solver uses the parallel planning interface with a single pipeline. The previously configured default stopping criteria was met by that planner (which doesn't have any effect since it is a single pipeline anyways). I updated the default, so this doesn't happen anymore

henningkayser commented on core/include/moveit/task_constructor/solvers/pipeline_planner.h:
```diff
@@ -56,48 +59,132 @@ MOVEIT_CLASS_FORWARD(PipelinePlanner);
+	PipelinePlanner(const rclcpp::Node::SharedPtr& node, const std::string& pipeline_name = "ompl",
+	                const std::string& planner_id = "");
+
+	[[deprecated("Deprecated: Use new constructor implementations.")]]  // clang-format off
```
Would it make sense to extend from PipelinePlanner instead of deprecating the old interface?

henningkayser commented on core/include/moveit/task_constructor/solvers/pipeline_planner.h:
```diff
@@ -56,48 +59,132 @@ MOVEIT_CLASS_FORWARD(PipelinePlanner);
+	 * \param [in] solution_selection_function Callback function to choose the best solution when multiple pipelines are used
+	 */
+	PipelinePlanner(const rclcpp::Node::SharedPtr& node,
+	                const std::unordered_map<std::string, std::string>& pipeline_id_planner_id_map,
```
usually, pipelines have default algorithms configured, right? Shouldn't the planner_id map specified per stage, if at all? The properties would be a great place to specify planner ids

sjahr replied:
I think the planner_id map should be defined because rather than using unconsciously some default algorithm, I think it should be explicitly configured which one is used.
To have an additional property to override this pipeline_id_planner_id_map on the stage level is a good idea. With that it would be possible to use a subset of the defined planner-pipeline pairs. Would this require more work in the stage implementations? If yes, I think this should be follow-up work.

sjahr replied:
I think the last commit achieves what you requested:
https://github.com/ros-planning/moveit_task_constructor/pull/450/commits/f57e65932e9f61741db99176800879de3291d218

henningkayser commented on core/src/solvers/pipeline_planner.cpp:
```diff
@@ -49,164 +49,176 @@
 #include <tf2_eigen/tf2_eigen.h>
 #endif
 
+namespace {
```
Unused? Please remove

sea-bass reviewed (commented):
I noticed these warnings when building `moveit_task_constructor_demos` with the latest version of MoveIt 2.

```
#39 1131.7 --- stderr: moveit_task_constructor_demo
#39 1131.7 /opt/underlay_ws/src/moveit_task_constructor/demo/src/fallbacks_move_to.cpp: In lambda function:
#39 1131.7 /opt/underlay_ws/src/moveit_task_constructor/demo/src/fallbacks_move_to.cpp:35:33: warning: ‘void moveit::task_constructor::solvers::PipelinePlanner::setPlannerId(const string&)’ is deprecated: Replaced with setPlannerId(pipeline_name, planner_id) [-Wdeprecated-declarations]
#39 1131.7    35 |                 pp->setPlannerId("PTP");
#39 1131.7       |                 ~~~~~~~~~~~~~~~~^~~~~~~
#39 1131.7 In file included from /opt/underlay_ws/src/moveit_task_constructor/demo/src/fallbacks_move_to.cpp:9:
#39 1131.7 /opt/underlay_ws/install/moveit_task_constructor_core/include/moveit/task_constructor/solvers/pipeline_planner.h:91:14: note: declared here
#39 1131.7    91 |         void setPlannerId(const std::string& /*planner*/) { /* Do nothing */
#39 1131.7       |              ^~~~~~~~~~~~
#39 1131.7 /opt/underlay_ws/src/moveit_task_constructor/demo/src/fallbacks_move_to.cpp: In lambda function:
#39 1131.7 /opt/underlay_ws/src/moveit_task_constructor/demo/src/fallbacks_move_to.cpp:41:33: warning: ‘void moveit::task_constructor::solvers::PipelinePlanner::setPlannerId(const string&)’ is deprecated: Replaced with setPlannerId(pipeline_name, planner_id) [-Wdeprecated-declarations]
#39 1131.7    41 |                 pp->setPlannerId("RRTConnect");
#39 1131.7       |                 ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
#39 1131.7 In file included from /opt/underlay_ws/src/moveit_task_constructor/demo/src/fallbacks_move_to.cpp:9:
#39 1131.7 /opt/underlay_ws/install/moveit_task_constructor_core/include/moveit/task_constructor/solvers/pipeline_planner.h:91:14: note: declared here
#39 1131.7    91 |         void setPlannerId(const std::string& /*planner*/) { /* Do nothing */
#39 1131.7       |              ^~~~~~~~~~~~
```

So I think this PR should also update those accordingly.

sjahr commented:
Ping @rhaschke @henningkayser

MarqRazz commented on core/include/moveit/task_constructor/solvers/pipeline_planner.h:
```diff
@@ -56,48 +59,132 @@ MOVEIT_CLASS_FORWARD(PipelinePlanner);
+	 */
+	bool setPlannerId(const std::string& pipeline_name, const std::string& planner_id);
+
+	/** \brief Set stopping criterion function for parallel planning
+	 * \param [in] stopping_criterion_callback New stopping criterion function that will be used
+	*/
+	void setStoppingCriterionFunction(const moveit::planning_pipeline_interfaces::StoppingCriterionFunction& stopping_criterion_callback);
+
+	/** \brief Set solution selection function for parallel planning
+	 * \param [in] solution_selection_function New solution selection that will be used
+	*/
+	void setSolutionSelectionFunction(const moveit::planning_pipeline_interfaces::SolutionSelectionFunction& solution_selection_function);
```
These two functions are explicitly for Parallel planning. If someone set them on a single planning pipeline what would happen? Are they just ignored? Does the user get a warning?

Should we consider adding to the function name to be explicit? `setParallelPlanningStoppingCriterionFunction`

sjahr replied:
No, the user doesn't get a warning but nothing happens. If only a single planner is used, no solution needs to be selected and no stopping decision based on other planning pipelines needs to be taken. I'd keep it that way

MarqRazz commented on core/include/moveit/task_constructor/solvers/pipeline_planner.h:
```diff
@@ -56,48 +59,132 @@ MOVEIT_CLASS_FORWARD(PipelinePlanner);
+	 * \param [in] from Start planning scene
+	 * \param [in] to Goal planning scene (used to create goal constraints)
+	 * \param [in] joint_model_group Group of joints for which this trajectory is created
+	 * \param [in] timeout ?
```
I assume this is in seconds.

Should we note this is the max planning time for the individual stage and not the complete task. That is if there were N connecting parent stages then the max timeout would be `N*timeout`

MarqRazz commented on core/include/moveit/task_constructor/solvers/pipeline_planner.h:
```diff
@@ -56,48 +59,132 @@ MOVEIT_CLASS_FORWARD(PipelinePlanner);
-	planning_pipeline::PlanningPipelinePtr planner_;
+	/** \brief Function that actually uses the planning pipelines to solve the given planning problem. It is called by
+	 * the public plan function after the goal constraints are generated. This function uses a predefined number of
+	 * planning pipelines in parallel to solve the planning problem and choose the automatically the best (user-defined)
```
Suggest removing `automatically`

`choose the best (user-defined) solution`

MarqRazz commented on core/src/solvers/pipeline_planner.cpp:
```diff
@@ -53,160 +53,175 @@ namespace moveit {
-		entry = planner;
+bool PipelinePlanner::setPlannerId(const std::string& pipeline_name, const std::string& planner_id) {
+	// Only set ID if pipeline exists. It is not possible to create new pipelines with this command.
+	if (pipeline_id_planner_id_map_.count(pipeline_name) > 0) {
+		pipeline_id_planner_id_map_[pipeline_name] = planner_id;
 	}
-	return planner;
+	RCLCPP_ERROR(node_->get_logger(),
+	             "PipelinePlanner does not have a pipeline called '%s'. Cannot set pipeline ID '%s'",
+	             pipeline_name.c_str(), planner_id.c_str());
+	return false;
```
Is this function reads like its missing a `return true` in the if statement?

sjahr replied:
Oh, yes!

MarqRazz reviewed (commented):
@sjahr what else can I do to help you get this across the finish line?

MarqRazz reviewed (approved):
After we decide what to do about the failing CI I think this is good to go. I have tested this with simulation and hardware with multiple robot configurations (6 and 7 DOF) and it works great!

sea-bass reviewed (approved):
LGTM

I still think we should find a way to ensure that PRs to the `humble` branch run humble CI, but we can handle that in a follow-on PR.

sjahr commented:
@rhaschke All that is blocking this PR from being merged, is that CI doesn't pass due to low code coverage. I just added a unit test for my changes and it barely increased coverage. Do you have any idea what is missing and/or if this is a problem caused by this PR?

sjahr commented:
It looks like the diff that is analyzed by code coverage is way bigger than this PR https://app.codecov.io/gh/ros-planning/moveit_task_constructor/pull/450?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=ros-planning#ed2ff5abe724fdee9649d476ae045460-R76
the reference head is https://app.codecov.io/gh/ros-planning/moveit_task_constructor/commit/ede7cb71cc302932b6fe427c714b84fa725aeae2 and not the latest of ros2 :thinking:

sjahr commented:
The ros2 branch in codev is out-of-date  https://app.codecov.io/gh/ros-planning/moveit_task_constructor/tree/ros2

sjahr commented:
> The ros2 branch in codev is out-of-date https://app.codecov.io/gh/ros-planning/moveit_task_constructor/tree/ros2

Looks like re-triggering the CI job for the ros2 branch fixed it
